### PR TITLE
5597 - add a module allows testing a big voters for a candidate

### PR DESCRIPTION
**Pull request by marknguyen85**

### CODE REVIEW

File: cmd/stress_sc/stress_sc.go [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+				tx, err := contractClient.Vote()
+				if err != nil {
+					logger.Errorw("failed to vote for candidate", "number", (i + 1), "error", err)
+				} else {
+					wErr := txutil.CheckTransStatus(contractClient.Client, tx)
+					if wErr != nil {
+						logger.Errorw("failed to checks the voting to candidate", "number", (i + 1), "account", addr, "error", wErr)
+					} else {
+						numberSuccess = numberSuccess + 1
+						totalTime = totalTime + time.Since(start).Milliseconds()
```
Comment (mr-nmh) [inline cmd/stress_sc/stress_sc.go]: go 1.12 doesn't have `.Milliseconds()`

File: Makefile [hunk 2]
```diff
@@ -23,4 +23,9 @@ blockmonitor:
 stakingcontract:
```
Comment (mr-nmh) [inline Makefile:23]: You must add `stakingcontract` to `.PHONY:` at line 1
If not, you will get `make: `stakingcontract' is up to date.`

File: Makefile [hunk 3]
```diff
@@ -23,4 +23,9 @@ blockmonitor:
...
 	go build -v -o ./build/sc ./cmd/stakingcontract
 	@echo "Done building."
-	@echo 'Run "./build/sc" to interact with staking contract.'
\ No newline at end of file
+	@echo 'Run "./build/sc" to interact with staking contract.'
+
+stresssc:
+	go build -v -o ./build/stresssc ./cmd/stress_sc
+	@echo "Done building."
```
Comment (mr-nmh) [inline Makefile:30]: Why do we need @ before `echo`?

File: cmd/stress_sc/main.go [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+	"github.com/urfave/cli"
+
+	"github.com/evrynet-official/evrynet-tools/lib/node"
+	sc "github.com/evrynet-official/evrynet-tools/stakingcontract"
+)
+
+func main() {
+	app := cli.NewApp()
+	app.Name = "stress-test tool"
+	app.Usage = "testing for staking contract"
```
Comment (mr-nmh) [inline cmd/stress_sc/main.go]: Should be `Stress test for staking contract`

File: cmd/stress_sc/stress_sc.go [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+				nonce, err := contractClient.Client.PendingNonceAt(context.Background(), addr)
+				if err != nil {
+					return err
+				}
+
+				optTrans := bind.NewKeyedTransactor(voterPk)
+				optTrans.Nonce = new(big.Int).SetUint64(nonce)
+				contractClient.TranOps = optTrans
+
+				logger.Infow("begin vote for candidate", "number", (i + 1), "account", addr, "amount", contractClient.Amount)
```
Comment (mr-nmh) [inline cmd/stress_sc/stress_sc.go]: Redundant parentheses at (i + 1), please remove it.

File: cmd/stress_sc/stress_sc.go [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+				start := time.Now()
+				tx, err := contractClient.Vote()
+				if err != nil {
+					logger.Errorw("failed to vote for candidate", "number", (i + 1), "error", err)
```
Comment (mr-nmh) [inline cmd/stress_sc/stress_sc.go]: didto^

File: cmd/stress_sc/stress_sc.go [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+				} else {
+					wErr := txutil.CheckTransStatus(contractClient.Client, tx)
+					if wErr != nil {
+						logger.Errorw("failed to checks the voting to candidate", "number", (i + 1), "account", addr, "error", wErr)
```
Comment (mr-nmh) [inline cmd/stress_sc/stress_sc.go]: didto

File: stakingcontract/sc.go [hunk 8]
```diff
@@ -36,13 +36,24 @@ var (
 	gasLimitFlag = cli.Uint64Flag{
 		Name:  "gaslimit",
 		Usage: "the gaslimit to execute the call to the contract.",
-		Value: 8000000,
+		Value: 0,
```
Comment (mr-nmh) [inline stakingcontract/sc.go]: Why did you change it?

File: stakingcontract/sc.go [hunk 9]
```diff
@@ -198,6 +228,15 @@ func (c ContractClient) GetVoters(opts *bind.CallOpts) ([]common.Address, error)
 	return response, nil
 }
 
+// GetVoterStake returns the staking of a voter from SC
+func (c ContractClient) GetVoterStake(opts *bind.CallOpts, voter common.Address) (*big.Int, error) {
```
Comment (mr-nmh) [inline stakingcontract/sc.go]: Should use pointer receiver func `func (c *ContractClient)`
Please update for another function.

File: stakingcontract/sc.go [hunk 10]
```diff
@@ -206,3 +245,12 @@ func (c ContractClient) getCandidateData(opts *bind.CallOpts) (bool, common.Addr
 	}
 	return response.IsActiveCandidate, response.Owner, response.TotalStake, nil
 }
+
+// GetVoterStakes returns the staking of a voter from SC
+func (c ContractClient) GetVoterStakes(opts *bind.CallOpts, voters []common.Address) ([]*big.Int, error) {
```
Comment (mr-nmh) [inline stakingcontract/sc.go]: didto^

File: stakingcontract/sc_test.go [hunk 11]
```diff
@@ -61,6 +57,13 @@ func TestContractClient(t *testing.T) {
 		TranOps:   bind.NewKeyedTransactor(senderPk),
 		Logger:    zap,
 	}
+}
+
+func TestContractClient(t *testing.T) {
+	var (
+		candidate = common.HexToAddress(Candidate)
+	)
+	contractClient := initContractClient(candidate)
```
Comment (mr-nmh) [inline stakingcontract/sc_test.go]: Group this variable

File: stakingcontract/sc_test.go [hunk 12]
```diff
@@ -74,10 +77,10 @@ func TestContractClient(t *testing.T) {
 		t.Errorf("Register() error = %v", err)
 	}
 
-	if err = txutil.CheckTransStatus(client, tx); err != nil {
+	if err = txutil.CheckTransStatus(contractClient.Client, tx); err != nil {
 		t.Error("can not register new candidate", err)
 	} else {
-		time.Sleep(EpochTime * time.Second)
+		//time.Sleep(EpochTime * time.Second)
```
Comment (mr-nmh) [inline stakingcontract/sc_test.go]: plz remove it

**Review by mr-nmh — CHANGES_REQUESTED** [review]
I see we still use "github.com/evrynet-official/evrynet-client/ethclient" at `blockmonitor/blockchain.go` please change it to "github.com/Evrynetlabs/evrynet-node/evrclient"

File: Makefile (hunk 3 above)
Comment (marknguyen85) [reply Makefile:30]: its same with syntax above, however i will remove it if necessary.

File: stakingcontract/sc.go (hunk 8 above)
Comment (marknguyen85) [reply stakingcontract/sc.go]: ah mistake, i've just updated for it.

File: stakingcontract/sc.go (hunk 9 above)
Comment (marknguyen85) [reply stakingcontract/sc.go]: ok updated

File: Makefile (hunk 2 above)
Comment (marknguyen85) [reply Makefile:23]: yes, i've added

File: Makefile (hunk 3 above)
Comment (mr-nmh) [reply Makefile:30]: It's redundant, plz remove it

Comment (marknguyen85) [reply Makefile:30]: If we remove it, the result will print on the screen including the entire command. 
For exam: 
- `echo "Done building."` the result is:

```
echo "Done building."
Done building.
```
- else if using `@echo "Done building."` the result is:

```
Done building.
```
so i think we should not remote it.

Comment (mr-nmh) [reply Makefile:30]: ok

Comment (mr-nmh) [reply Makefile:30]: oh, I see

**Review by cyberliem — APPROVED** [review]
